apps/sync/js/adapters/history.js

```diff
@@ -0,0 +1,285 @@
+  operations.
+
+    History records are stored locally in a DataStore with format [1] while
+  history records coming from Firefox Sync (via Kinto collection) have
+  format[2]. We need to convert from [1] to [2] and viceversa. Also, we need to
+  add the remote record ID[1.1] from the Kinto collection to its corresponding
+  match in the local DataStore, so we can remove the local record afterwards
+  when any deleting record requests with the format [3] are coming from FxSync.
+
+  [1] Records stored in Places DataStore (PDS): {
+    "url": "http://mozilla.org/", // KEY in PDS
+    "title": "Mozilla",
+    "icons": {
+      "http://mozilla.org/favicon.ico": {
+        "sizes": []
+      }
+    },
+    "frecency": 1,
+    "visits": [
+      // NOTICE: date/time without ms
+      1442247252490, 1442247250001
+    ],
+    "screenshot": {},
+    "fxsyncId" "REMOTE_ID", // [1.1]
+    "visited": 1442247252490
+  }
+
+  [2] Add/Update Records from History Collection (HC): {
+    "id": "zMgfGkRinh92",
+    "sortindex": 2000,
+    "last_modified": 1442247272150,
+    "payload": {
+      "id": "zMgfGkRinh92",
+      "histUri": "http://mozilla.org/",
+      "title": "Mozilla",
+      "visits": [
+         // NOTICE: date/time with ms
+        { "date": 1442247252490018, "type": 2 },
+        { "date": 1442247250001234, "type": 2 }
+      ]
+    },
+    "_status": "synced"
+  }
+
+  [3] Delete Records from History Collection (HC): {
+    "id": "_Avscjx5srFy",
+    "sortindex": 100,
+    "last_modified": 1441985077970,
+    "payload": {
+      "id": "_Avscjx5srFy",
+      "deleted": true
+    },
+    "_status": "synced"
+  }
+**/
+  _update(remoteRecords, lastModifiedTime) {
+    var places = [];
+    for (var i = 0; i < remoteRecords.length; i++) {
+      var payload = remoteRecords[i].payload;
+      if (!Number.isInteger(remoteRecords[i].last_modified)) {
+        console.warn('Incorrect payload::last_modified? ', payload);
+        continue;
+      }
+      if (remoteRecords[i].last_modified <= lastModifiedTime) {
+        break;
+      }
+      if (payload.deleted) {
+        places.push({
+          deleted: true,
+          fxsyncId: payload.id
+        });
+        continue;
+      }
+      if (!payload.histUri || !payload.visits || !payload.visits.length) {
+        console.warn('Incorrect payload? ', payload);
+        continue;
+      }
+
+      places.push({
+        url: payload.histUri,
+        title: payload.title,
+        visits: payload.visits.map(elem => Math.floor(elem.date / 1000)),
+        fxsyncId: payload.id
+      });
+    }
+
+    if (places.length === 0) {
+      return Promise.resolve(false);
+    }
+
+    return HistoryHelper.updatePlaces(places).then(() => {
+      var latestMtime = remoteRecords[0].last_modified;
+      return HistoryHelper.setSyncedCollectionMtime(latestMtime);
+    }).then(() => {
+      // Always return false for a read-only operation.
+      return Promise.resolve(false);
```

Comment:
@michielbdejong Could you help to explain here? In my memory, you gave me a suggestion that returning false to avoid the second round update when there is no any collection modification. Thanks.